src/cmdline_manager.ts

```diff
@@ -1,128 +1,208 @@
+import { calculateInputAfterTextChange } from "./utils/cmdline_text";
+import { GlyphChars } from "./constants";
+
+const logger = createLogger("CmdLine", false);
 
 export class CommandLineManager implements Disposable {
     private disposables: Disposable[] = [];
-    /**
-     * Simple command line UI
-     */
-    private commandLine?: CommandLineController;
-    /**
-     * Commandline timeout
-     */
-    private cmdlineTimer?: NodeJS.Timeout;
-
-    /**
-     * The last text typed in the UI, used to calculate changes
-     */
+
+    // The quickpick used to render the command line
+    private input: QuickPick<QuickPickItem>;
+
+    // The last text typed in the UI, used to calculate changes
     private lastTypedText: string = "";
 
-    private get client() {
-        return this.main.client;
-    }
+    // On suggestion selection, we don't want to send <CR> to nvim, so we ignore the accept event.
+    private ignoreAcceptEvent = false;
+    // On cmdline_hide, we close the quickpick. This flag is used to ignore that event so we don't send an <Esc> to nvim.
+    private ignoreHideEvent = false;
+
+    // When we type, we send updates to nvim. We want to ignore updates coming from nvim, because it may interfere with typing.
+    // However, bindings are expected to cause the cmdline content to change, so we use this flag to listen to those updates.
+    private redrawExpected = true;
+
+    // When updates come from nvim, we write to the input field.
+    // We don't want to send those updates back to nvim, so we use this counter to keep track of the number of onChange to ignore.
+    private pendingNvimUpdates = 0;
 
     public constructor(private main: MainController) {
         eventBus.on("redraw", this.handleRedraw, this, this.disposables);
+        this.input = window.createQuickPick();
+        (this.input as any).sortByLabel = false;
+        this.input.ignoreFocusOut = true;
+        this.input.buttons = [
+            {
+                iconPath: new ThemeIcon("close"),
+                tooltip: "Cancel",
+            },
+            {
+                iconPath: new ThemeIcon("check"),
+                tooltip: "Accept",
+            },
+        ];
+        this.reset();
+        this.disposables.push(
+            this.input,
+            this.input.onDidAccept(this.onAccept),
+            this.input.onDidChangeValue(this.onChange),
+            this.input.onDidHide(this.onHide),
+            this.input.onDidChangeSelection(this.onSelection),
+            this.input.onDidTriggerButton(this.onButton),
+            commands.registerCommand("vscode-neovim.commit-cmdline", this.onAccept),
+            commands.registerCommand("vscode-neovim.send-cmdline", this.sendRedraw),
+            commands.registerCommand("vscode-neovim.test-cmdline", this.testCmdline),
+        );
     }
 
     public dispose() {
-        this.commandLine?.dispose();
         disposeAll(this.disposables);
     }
 
+    private reset() {
```

Comment:
This feels brittle under maintenance; if someone adds a property, they'll have to remember to add it here. Might it be best to encapsulate this state in a class, and then upon the call to `reset`, just reconstruct a new object? something like

```ts
private reset () {
	this.cmdlineState = new CmdilneState();
	// and maybe still do the `this.input` resetting, since we can't encapsulate that
}
````